Reduce docker cache invalidation

Gehock opened the pull request:
This prevents invalidating the build cache accidentally when temporary files or backup files change in the worktree, or when git operations such as branch creation are performed.

~~gitignore is just for general cleanliness~~

Gehock committed:
Reduce docker cache invalidation

This prevents invalidating the build cache accidentally when temporary files or backup files change in the worktree, or when git operations such as branch creation are performed.

OverkillGuy commented:
Sounds very editor specific, hence user specific: have you considered making a user-wide `.gitignore` file with such patterns?
Docs say: `$XDG_CONFIG_HOME/git/ignore` is valid for such a thing.

Gehock commented:
The `.gitignore` changes were included mostly to just keep `.gitignore` and `.dockerignore` in sync with each other. They were mostly just a nice-to-have and can be replaced with a global `.gitignore`, so I removed them. As far as I can tell, Docker has no support for global ignore files (there's a [request](https://github.com/moby/moby/issues/12843) that has been open for the past 8 years).

I also added an entry for `.git`, which I had forgot earlier: at least for now the Dockerfiles don't do any git operations during the build process, so the `.git` directory isn't required in the build context and ignoring it reduces the chances of invalidating the cache unnecessarily.

Ignoring vim's swap files and emacs's backup files is quite editor-specific, I agree, but I'd argue that the two are common enough editors to warrant a change in an opinionated template (the template already includes ignores for fairly arbitrary paths like `ENV` and `venv.bak`, which seem to me a lot more user-specific than vim or emacs ;) ). Speaking from experience, accidentally busting docker's cache because of automatically generated editor files is a symptom that's not always easy to notice, especially if the files are ignored by git (so that they don't show up as changed in `git status`, while docker still sees the changes).

Gehock commented:
Moved the additions closer to the top of the file, plus rebased over current HEAD.
